README.md
# NOVA

Basic Vagrant VM for Node.js development using VirtualBox, Ubuntu 14.04, Nginx and MongoDB.

## What's included?

* [Ubuntu 14.04](https://www.ubuntu.com)
* [Nginx](https://nginx.org/en/)
* [Node.js](https://nodejs.org)
* [npm](https://www.npmjs.com/)
* [MongoDB](https://www.mongodb.com/)
* [PM2](https://github.com/Unitech/pm2)
* [MailDev](https://github.com/djfarrelly/MailDev)
* [Git](https://git-scm.com/)
* [curl](https://curl.haxx.se/)

## Getting Started

1. Download and install [VirtualBox](https://www.virtualbox.org/)
2. Download and install [Vagrant](https://www.vagrantup.com/)
   1. Install [vagrant-hostmanager](https://github.com/smdahlen/vagrant-hostmanager)
      `vagrant plugin install vagrant-hostmanager`
   2. (Optional) Install [vagrant-timezone](https://github.com/tmatilai/vagrant-timezone)
      `vagrant plugin install vagrant-timezone`
3. Clone this repository
    ```bash
    git clone https://github.com/attilabuti/nova.git
    ```
4. Start the VM
    ```bash
    cd nova
    vagrant up
    ```

You can now access your project at [http://projectname.local](http://projectname.local).

The web root is located in the project directory at `app/` and you can install your files there.

## Default VM parameters

You can customize settings by editing config.yaml file.

```yaml
name: projectname
hostname: projectname.local
aliases:
  - mail.projectname.local
virtualbox:
  gui: false
  cpus: 2
  memory: 1024
ip: 10.10.10.150
timezone: Europe/Budapest
appPort: 3000
mongodb:
  user: admin
  password: admin
```

## Usage

Some basic information on interacting with the vagrant box.

### Port Forwards

* 80 - Nginx
* 443 - Nginx
* 1025 - MailDev SMTP port
* 1080 - MailDev web interface
* 9229 - Node.js Inspector
* 27017 - MongoDB

### MongoDB

* Hostname: projectname.local
* Port: 27017
* User: admin
* Password: admin

**Note:** Remote database access is enabled by default, so you can access the MongoDB database using your favorite client with the above credentials.

## What is the provision.sh script doing?

* This file is only run when the virtual machine is initially created or recreated after a `vagrant destroy`
* Updates software on the VM
* Installs necessary packages
* Install Nginx, Node.js, MongoDB, MailDev, PM2
* Generate a self-signed SSL certificate
* Configure Nginx, MongoDB
* Starts Nginx, MongoDB, MailDev

## What is the startup.sh script doing?

This file is run every time the virtual machine is started from a `vagrant up`.

## MailDev

### What is MailDev?

[MailDev](https://github.com/djfarrelly/MailDev) is a simple way to test your project's generated emails during development with an easy to use web interface that runs on your machine built on top of Node.js.

### Using MailDev

Load [http://mail.projectname.local](http://mail.projectname.local) in your browser to view the MailDev interface.

## Vagrant

Vagrant is [very well documented](https://www.vagrantup.com/docs/index.html) but here are a few common commands:

* `vagrant up` starts the virtual machine and provisions it
* `vagrant suspend` will essentially put the machine to 'sleep' with `vagrant resume` waking it back up
* `vagrant halt` attempts a graceful shutdown of the machine and will need to be brought back with `vagrant up`
* `vagrant ssh` gives you shell access to the virtual machine
*	`vagrant destroy` stops and deletes all traces of the vagrant machine

## Changelog

### [1.0.0] (2017-10-20)
* Initial release

## Issues

Submit the [issues](https://github.com/attilabuti/nova/issues) if you find any bug or have any suggestion.

## Contribution

Fork the [repo](https://github.com/attilabuti/nova) and submit pull requests.

## License

This project is licensed under the [MIT License](https://raw.githubusercontent.com/attilabuti/nova/master/LICENSE).

CHANGELOG.md
# Change Log
All Notable changes to `nova` will be documented in this file.

### [1.0.0] (2017-10-20)
* Initial release

app/app.js
const fs = require('fs');
const http = require('http');
const path = require('path');
const spawn = require('child_process').spawnSync;

const port = 3000;

String.prototype.fmt = function(hash) {
    let string = this;

    for (key in hash) {
        string = string.replace(new RegExp('\\{{' + key + '\\}}', 'gm'), hash[key]);
    }

    return string;
}

http.createServer((req, res) => {
    fs.readFile(path.join(__dirname, 'app.html'), 'utf8', (err, data) => {
        data = data.toString().fmt({
            nodeVersion: spawn('node', ['-v']).stdout.toString().replace('v', ''),
            nginxVersion: spawn('nginx', ['-v']).stderr.toString().replace('nginx version: nginx/', ''),
            pm2Version: spawn('pm2', ['-v']).stdout.toString(),

            maildevVersion: spawn('maildev', ['--version']).stdout.toString(),
            maildevRunning: ((spawn('pm2', ['show', 'maildev']).stdout.toString().indexOf('online') > -1) ? 'check' : 'times'),

            mongoVersion: spawn('mongod', ['--version', '| grep "db version"']).stdout.toString().match(/\d+(\.\d+)+/)[0],
            mongoRunning: ((spawn('service', ['mongod', 'status']).stdout.toString().indexOf('start/running') > -1) ? 'check' : 'times'),

            host: req.headers.host
        });

        res.writeHead(200, { 'Content-Type': 'text/html', 'Content-Length': data.length });
        res.write(data);
        res.end();
    });
}).listen(port);

scripts/provision.sh
#!/usr/bin/env bash

# Exit on error
set -e

hostname=$1
appPort=$2
mongodbUser=$3
mongodbPass=$4

echo "--- Import the public key used by the package management system ---"
sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 0C49F3730359A14518585931BC711F9BA15703C6 # MongoDB

echo "--- Create a list file for MongoDB ---"
echo "deb [ arch=amd64 ] http://repo.mongodb.org/apt/ubuntu trusty/mongodb-org/3.4 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-3.4.list

echo "--- Add repo for latest stable Nginx ---"
sudo add-apt-repository -y ppa:nginx/stable

echo "--- Update package list ---"
sudo apt-get update

echo "--- Installing Base Packages ---"
sudo apt-get install -y curl git htop build-essential openssl libssl-dev

echo "--- Installing Nginx ---"
sudo apt-get -y install nginx

echo "--- Creating SSL certificate ---"
sudo mkdir /etc/nginx/ssl
sudo su -c "openssl req -newkey rsa:2048 -x509 -nodes -keyout /etc/nginx/ssl/$hostname.key -new -out /etc/nginx/ssl/$hostname.cert -subj /CN=$hostname -reqexts SAN -extensions SAN -config <(cat /etc/ssl/openssl.cnf <(printf '[SAN]\nsubjectAltName=DNS:$hostname')) -sha256 -days 3650"

echo "--- Configuring Nginx ---"
# Remove default sites from Nginx
sudo rm /etc/nginx/sites-enabled/default
sudo rm /etc/nginx/sites-available/default

echo "server {
    listen 80;

    server_name mail.$hostname;

    location / {
        proxy_pass http://localhost:1080;
    }
}" > /etc/nginx/sites-available/maildev

echo "server {
    listen 80;
    listen 443 default ssl;

    server_name $hostname *.$hostname;

    access_log /var/log/nginx/access.log;
    error_log /var/log/nginx/error.log;

    ssl_certificate /etc/nginx/ssl/$hostname.cert;
    ssl_certificate_key /etc/nginx/ssl/$hostname.key;

    location / {
        proxy_pass http://localhost:$appPort;
        proxy_http_version 1.1;
        proxy_set_header Upgrade \$http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host \$host;
        proxy_cache_bypass \$http_upgrade;
    }
}" > /etc/nginx/sites-available/$hostname

# Make a symbolic link
sudo ln -s /etc/nginx/sites-available/maildev /etc/nginx/sites-enabled/maildev
sudo ln -s /etc/nginx/sites-available/$hostname /etc/nginx/sites-enabled/$hostname

sudo service nginx reload

echo "--- Installing Node.js 8.x ---"
curl -sL https://deb.nodesource.com/setup_8.x | sudo -E bash -
sudo apt-get -y install nodejs

echo "--- Installing MongoDB ---"
sudo apt-get -y install mongodb-org

echo "--- Create MongoDB user ---"
mongo admin --eval "db.createUser({ user: '$mongodbUser', pwd: '$mongodbPass', roles: [ { role: 'root', db: 'admin' } ] });"

sudo service mongod stop
sudo sed -i 's/bindIp\: 127\.0\.0\.1/bindIp\: 0\.0\.0\.0/' /etc/mongod.conf
cat <<EOT >> /etc/mongod.conf
security:
  authorization: "enabled"
EOT
sudo service mongod start

echo "--- Installing PM2 ---"
sudo npm install -g pm2 --silent
sudo env PATH=$PATH:/usr/bin /usr/lib/node_modules/pm2/bin/pm2 startup upstart -u vagrant --hp /home/vagrant >/dev/null

echo "--- Installing MailDev ---"
sudo npm install -g maildev --silent

echo "--- Start MailDev ---"
sudo su vagrant -c "pm2 start maildev"
sudo su vagrant -c "pm2 save"

echo "--- All done! ---"